### PR TITLE
[SCB-2043] flow control support custom / bug fix

### PR DESCRIPTION
**Pull request by GuoYL123**
(cherry picked from commit eed0e46cf448b0caab1b7211f9780bbd82f55a6b)

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install -Pit` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java [hunk 1]
```diff
@@ -211,6 +212,8 @@ public QpsControllerManager setGlobalQpsStrategy(String globalLimitKey, String g
   private AbstractQpsStrategy chooseStrategy(String globalConfigKey, Long limit, Long bucket,
       String strategyKey) {
     AbstractQpsStrategy strategy;
+    AbstractQpsStrategy customStrategy = SPIServiceUtils
```
Comment (liubao68) [inline handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java:215]: Maybe this code can be better refactored.

1. all strategies are SPI 
2. any implementation have a name
3. choose the target strategy by configured name and default to 'FixedWindow'

Comment (GuoYL123) [reply handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java:215]: ` new TokenBucketStrategy(globalConfigKey, limit, bucket);`
`new FixedWindowStrategy(globalConfigKey, limit);`
has different parameters.

Comment (liubao68) [reply handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java:215]: Hidden parameters in implementation and I think will work.

Comment (GuoYL123) [reply handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java]: done

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java [hunk 2]
```diff
@@ -210,21 +209,17 @@ public QpsControllerManager setGlobalQpsStrategy(String globalLimitKey, String g
...
-      case LeakyBucket:
-        strategy = new LeakyBucketStrategy(globalConfigKey, limit);
-        break;
-      case TokenBucket:
-        strategy = new TokenBucketStrategy(globalConfigKey, limit, bucket);
-        break;
-      case SlidingWindow:
-      default:
-        strategy = new FixedWindowStrategy(globalConfigKey, limit);
+    AbstractQpsStrategy strategy = new FixedWindowStrategy(globalConfigKey, limit);
```
Comment (liubao68) [inline handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java:212]: strategyKey can have default value 'FixedWindow', wrong used defined value can throw exception. So here do not need new a default strategy.

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java [hunk 3]
```diff
@@ -210,21 +209,17 @@ public QpsControllerManager setGlobalQpsStrategy(String globalLimitKey, String g
...
+    List<AbstractQpsStrategy> customStrategy = SPIServiceUtils
+        .getOrLoadSortedService(AbstractQpsStrategy.class);
+    for (AbstractQpsStrategy abstractQpsStrategy : customStrategy) {
+      if (abstractQpsStrategy.name().equals(strategyKey)) {
+        strategy = abstractQpsStrategy;
```
Comment (liubao68) [inline handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java:217]: Add a break is better.

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java [hunk 4]
```diff
@@ -210,21 +209,17 @@ public QpsControllerManager setGlobalQpsStrategy(String globalLimitKey, String g
...
-        strategy = new TokenBucketStrategy(globalConfigKey, limit, bucket);
-        break;
-      case SlidingWindow:
-      default:
-        strategy = new FixedWindowStrategy(globalConfigKey, limit);
+    AbstractQpsStrategy strategy = new FixedWindowStrategy(globalConfigKey, limit);
+    List<AbstractQpsStrategy> customStrategy = SPIServiceUtils
+        .getOrLoadSortedService(AbstractQpsStrategy.class);
+    for (AbstractQpsStrategy abstractQpsStrategy : customStrategy) {
+      if (abstractQpsStrategy.name().equals(strategyKey)) {
```
Comment (liubao68) [inline handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java:216]: strategyKey  better to rename to 'strategy', it is not a key , but a value.

Comment (GuoYL123) [reply handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java]: already has local value 'strategy'

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java (hunk 2 above)
Comment (GuoYL123) [reply handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java]: throw a Exception

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java [hunk 5]
```diff
@@ -209,22 +208,21 @@ public QpsControllerManager setGlobalQpsStrategy(String globalLimitKey, String g
...
-      case LeakyBucket:
-        strategy = new LeakyBucketStrategy(globalConfigKey, limit);
-        break;
-      case TokenBucket:
-        strategy = new TokenBucketStrategy(globalConfigKey, limit, bucket);
-        break;
-      case SlidingWindow:
-      default:
-        strategy = new FixedWindowStrategy(globalConfigKey, limit);
+      String strategyName) {
```
Comment (liubao68) [inline handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java:211]: strategyName do not have a default value.

Comment (GuoYL123) [reply handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java:211]: when name is incorrect ， throw a Exception ?  
`    if (strategy == null) {
      throw new ServiceCombException("the qps strategy name is not exist , please check.");
    }`

Comment (GuoYL123) [reply handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java:212]: done

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java (hunk 3 above)
Comment (GuoYL123) [reply handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java]: done.

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java [hunk 6]
```diff
@@ -17,20 +17,20 @@
...
 import org.apache.servicecomb.qps.strategy.AbstractQpsStrategy;
-import org.apache.servicecomb.qps.strategy.FixedWindowStrategy;
-import org.apache.servicecomb.qps.strategy.LeakyBucketStrategy;
-import org.apache.servicecomb.qps.strategy.StrategyType;
-import org.apache.servicecomb.qps.strategy.TokenBucketStrategy;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import com.netflix.config.DynamicProperty;
+import org.springframework.util.StringUtils;
```
Comment (liubao68) [inline handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java:33]: use commons lang3 api

Comment (GuoYL123) [reply handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java]: done.

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java [hunk 7]
```diff
@@ -209,22 +210,25 @@ public QpsControllerManager setGlobalQpsStrategy(String globalLimitKey, String g
...
+      strategy = strategyFactory.getStrategy(strategyName);
+      if (strategy != null) {
         break;
-      case SlidingWindow:
-      default:
-        strategy = new FixedWindowStrategy(globalConfigKey, limit);
+      }
+    }
+    if (strategy == null) {
+      throw new ServiceCombException("the qps strategy name is not exist , please check.");
```
Comment (liubao68) [inline handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java:227]: exeception message is better to include `strategyName`

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/strategy/DefaultStrategyFactory.java [hunk 8]
```diff
@@ -14,27 +14,20 @@
...
-    if (StringUtils.isEmpty(type)) {
-      return StrategyType.FixedWindow;
-    }
+public class DefaultStrategyFactory implements IStrategyFactory {
 
-    try {
-      return StrategyType.valueOf(type.toUpperCase());
-    } catch (IllegalArgumentException e) {
-      return StrategyType.FixedWindow;
+  public AbstractQpsStrategy getStrategy(String strategyName) {
```
Comment (liubao68) [inline handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/strategy/DefaultStrategyFactory.java:21]: It's better to use `createStrategy`

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java (hunk 7 above)
Comment (GuoYL123) [reply handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/QpsControllerManager.java]: done

File: handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/strategy/DefaultStrategyFactory.java (hunk 8 above)
Comment (GuoYL123) [reply handlers/handler-flowcontrol-qps/src/main/java/org/apache/servicecomb/qps/strategy/DefaultStrategyFactory.java]: done

**Pull request merged**